src/main/kotlin/com/openlattice/indexing/BackgroundExpiredDataDeletionService.kt

```diff
@@ -0,0 +1,250 @@
+                            w.elapsed(TimeUnit.MILLISECONDS)
+                    )
+                } else {
+                    logger.info("Skipping expired data deletion as it is not enabled.")
+                }
+            } finally {
+                taskLock.unlock()
+            }
+        } else {
+            logger.info("Not starting new expired data deletion job as an existing one is running.")
+        }
+    }
+
+    //deletes expired data from data and ids table in postgres and elasticsearch
+    private fun deleteExpiredData(entitySet: EntitySet): Int {
+        logger.info(
+                "Starting deletion of expired data for entity set {} with id {}",
+                entitySet.name,
+                entitySet.id
+        )
+        var dataTableDeleteCount = 0
+        if (entitySet.expiration != null) {
+            val pair: Pair<Set<UUID>, Int> = deleteExpiredDataFromDataTable(entitySet.id, entitySet.expiration)
+
+            dataTableDeleteCount = pair.second
+            if (dataTableDeleteCount <= 0) {
+                logger.info("Entity set {} has no expired data", entitySet.name)
+                return dataTableDeleteCount
+            }
+            val expiredEntityKeyIds = pair.first
+            var elasticsearchDataDeleted = false
+            var idsTableDeleteCount = 0
+            if (expiredEntityKeyIds.isNotEmpty()) {
+                elasticsearchDataDeleted = deleteExpiredDataFromElasticSearch(entitySet.id, entitySet.entityTypeId, expiredEntityKeyIds)
+                idsTableDeleteCount = deleteExpiredDataFromIdsTable(expiredEntityKeyIds)
+            }
+
+            //compare deletion from data table and ids table and whether data was deleted from elasticsearch
+            check(expiredEntityKeyIds.size == idsTableDeleteCount) { "Number of entities deleted from data and ids table are not the same. UH OH." } //do something better
+            check(elasticsearchDataDeleted) { "Expired data not deleted from elasticsearch. UH OH." } // also do something better
+            logger.info("Completed deleting {} expired elements from entity set {}.",
+                    idsTableDeleteCount,
+                    entitySet.name)
+        } else {
+            logger.info(
+                    "Entity set {} does not have a data expiration policy.",
+                    entitySet.name
+            )
+        }
+        return dataTableDeleteCount
+    }
+
+    private fun deleteExpiredDataFromDataTable(entitySetId: UUID, expiration: DataExpiration): Pair<Set<UUID>, Int> {
+        val connection = hds.connection
+        val comparisonField: String
+        val expirationField: Any
+        val expirationFieldSQLType: Int
+        var deleteCount = 0;
+        when (expiration.expirationFlag) {
+            ExpirationType.DATE_PROPERTY -> {
+                val propertyTypeId: UUID = expiration.startDateProperty.get()
+                val propertyType = propertyTypes[propertyTypeId]
```

Comment:
Reading from hazelcast maps is relatively expensive, so we should reduce/bulk these calls as much as we can. It would probably improve perf to load this data once per entity set and pass the datatype through to this function instead of loading it for every batch. (Fastest would be to map from entity sets -> expiration property types -> bulk project to datatypes to avoid serializing property type fields you don't care about)

Reply:
i believe this issue is solved in new version